File: python/binary_answer.py
def first_true(L: int, R: int, f):
    R += 1
    while L < R:
        mid = L+(R-L)//2
        if f(mid):
            R = mid
        else:
            L = mid+1
    return L


def last_true(L: int, R: int, f):
    L -= 1
    while L < R:
        mid = L+(R-L+1)//2
        if f(mid):
            L = mid
        else:
            R = mid-1
    return L


def test():
    a = list(range(10))

    def even(i):
        return a[i] % 2 == 0

    print(first_true(0, 10, even), last_true(0, 10, even))


test()

File: cpp/binary_answer.cpp
#include <functional>
template <typename T>
T first_true(T l, T r, std::function<bool(T)> &&f) //[l,r]
{
    r++;
    while (l < r)
    {
        T mid = l + (r - l) / 2;
        if (f(mid))
        {
            r = mid;
        }
        else
        {
            l = mid + 1;
        }
    }
    return l;
}

template <typename T>
T last_true(T l, T r, std::function<bool(T)> &&f) //[l,r]
{
    l--;
    while (l < r)
    {
        T mid = l + (r - l + 1) / 2;
        if (f(mid))
        {
            l = mid;
        }
        else
        {
            r = mid - 1;
        }
    }
    return l;
}
